test/cpu_reduce_util.hpp

```diff
@@ -29,19 +29,196 @@
 #include <half.hpp>
 #include <limits>
 #include <cmath>
+#include <cassert>
 #include <miopen/miopen.h>
 #include <miopen/reduce_common.hpp>
 
 #include <miopen/bfloat16.hpp>
 
 namespace reduce {
 
+template <typename T>
+static inline bool IsNan(T x)
+{
+    // C++ isnan() is used for float and double
+    return (std::isnan(x));
+};
```

Comment:
Cant we just do `using std::isnan;` and then call `isnan` unqualified? Then the half overload should be found by ADL lookup

Reply:
Do you mean std::isnan() can also be directed to call half's overloaded isnan() implementation ?    Can you point to me some document about which C++ interfaces can be used for non-builtin object like half_float::half, but which implements overloaded common interfaces like operator == and isnan(), isfinite()

Reply:
Also for static_cast<>， can it be called directly to do transformation between float/double/in and half_float::half ?

Reply:
Yes.  std::isnan() works for half_float::half.  But I kept the definition of IsNan() there in case other types are involved.

Reply:
> But I kept the definition of IsNan() there in case other types are involved.

This is overdesign.

> std::isnan() works for half_float::half.

`std::isnan()` can't find `isnan` for half, because the latter is defined in the `half_float` namespace.

> Can you point to me some document about which C++ interfaces can be used for non-builtin object like half_float::half,...

Please refer to https://en.cppreference.com/w/cpp/language/adl which is applicable to the design proposed by Paul.

Or you can import `isnan` for half into `std` like this
```C++
namespace std {
using half_float::isnan;
}
```
But I am not sure if this is good practice.

Reply:
You shouldn't call `std::isnan`. Instead just call `isnan` unqualified(without `std::`). This will lookup the function using ADL, which means the function will first check in the data types' associated namespace, otherwise it will fallback on the current namespace.

This means that calling `isnan` using `half` will lookup the function in the `half_float` namespace. However, for builtin types there is no associated namespace so you need to add `using std::isnan` to your current namespace so it will lookup `std::isnan` for builtin types. This is also what is recommended from [half's documentation](http://half.sourceforge.net/index.html#usage):

> Additionally the half_float namespace also defines half-precision versions for all mathematical functions of the C++ standard library, which can be used directly through ADL:

You should use the standard floating-point functions, which makes everything consistent. If we add a data type in the future then that data type should provide overloads for these functions that can be found with ADL as well.

Reply:
Understand.  Changed